### ISSUE TITLE
Hash algorithm in `FutexBucketVec.get_bucket()` leads to poor key distribution

### ISSUE THREAD

**Issue by tangruize** [asterinas/asterinas#1641]
### Current Feature

The bucket index is calculated by `((addr >> 2) / size) & (size - 1)`, where `size` is calculated by `((1 << 8) * num_cpus()).next_power_of_two()`. `addr` is typically obtained via `mmap()` syscall, which is currently not randomized. As a result, only a few buckets are used, leaving many unused. Even when virtual space is randomized, users still can disable `randomize_va_space`. This issue is exacerbated as the number of CPUs increases. For example, with 33 CPUs, `size` becomes `16384`, which leads to the addr shifting `2+14` bits, causing only one bucket to be used in most cases.

https://github.com/asterinas/asterinas/blob/05ff4415772bbf32a12a5cf49ba869beb4d70101/kernel/src/process/posix_thread/futex.rs#L195-L203

[Example](https://man7.org/linux/man-pages/man2/futex.2.html#EXAMPLES):

<details><summary>Click to show print patch</summary>

```diff
diff --git a/kernel/src/process/posix_thread/futex.rs b/kernel/src/process/posix_thread/futex.rs
index 26af862c..0e992533 100644
--- a/kernel/src/process/posix_thread/futex.rs
+++ b/kernel/src/process/posix_thread/futex.rs
@@ -4,8 +4,7 @@

 use intrusive_collections::{intrusive_adapter, LinkedList, LinkedListAtomicLink};
 use ostd::{
-    cpu::num_cpus,
-    sync::{Waiter, Waker},
+    cpu::num_cpus, early_println, sync::{Waiter, Waker}
 };
 use spin::Once;

@@ -199,6 +198,7 @@ impl FutexBucketVec {
             // simple hash
             addr / self.size()
         };
+        early_println!("addr: {:x}, index: {}, size: {}", key.addr(), index, self.size());
         (index, self.vec[index].clone())
     }

```

</details>

<details><summary>Click to show results</summary>

```txt
# make run SMP=33
/test/futex # ./futex-demo > out
addr: 7ffffffe3000, index: 16382, size: 16384
addr: 7ffffffe3004, index: 16382, size: 16384
addr: 7ffffffe3000, index: 16382, size: 16384
addr: 7ffffffe3004, index: 16382, size: 16384
addr: 7ffffffe3000, index: 16382, size: 16384
addr: 7ffffffe3004, index: 16382, size: 16384
addr: 7ffffffe3000, index: 16382, size: 16384
addr: 7ffffffe3004, index: 16382, size: 16384
addr: 7ffffffe3000, index: 16382, size: 16384
addr: 7ffffffe3004, index: 16382, size: 16384
addr: 7ffffffe3000, index: 16382, size: 16384
addr: 7ffffffe3004, index: 16382, size: 16384
addr: 7ffffffe3000, index: 16382, size: 16384
addr: 7ffffffe3004, index: 16382, size: 16384
addr: 7ffffffe3000, index: 16382, size: 16384
addr: 7ffffffe3004, index: 16382, size: 16384
addr: 7ffffffe3000, index: 16382, size: 16384
addr: 7ffffffe3004, index: 16382, size: 16384
addr: 7ffffffe3000, index: 16382, size: 16384
addr: 7ffffffe3004, index: 16382, size: 16384
addr: 7ffffffe3000, index: 16382, size: 16384
addr: 7ffffffe3004, index: 16382, size: 16384
addr: 7ffffffe3000, index: 16382, size: 16384
addr: 7ffffffe3000, index: 16382, size: 16384
addr: 7ffffffe3004, index: 16382, size: 16384
```

</details>

### Enhancement Description

Simply combining the `key.addr` and `key.pid` fields with XOR maybe better.

**Comment by tatetian** [comment]
This is indeed a problem if every program is written in the way shown by the example. But lucky for us, real programs are not written that way. Futexes are provided by Linux to implement highly-efficient, synchronization primitives such as mutexes. The memory of mutex objects are allocated from the heap managed by libc. Libc does this by obtaining pages from the kernel via `mmap` and then logically divides the pages into fixed-size slots, each of which can store an object. So the addresses of these objects (modulo the page size) will be evenly distributed.

**Comment by tangruize** [comment]
Thank you for the explanation.

While this assumption may hold within a single process, it may not be as reliable in multi-process scenarios where virtual address space is not randomized. In such cases, multiple processes may map the same virtual address for `pthread_mutex_t` and `pthread_cond_t` objects, leading to increased contention, especially if many programs are using these synchronization primitives.

**Comment by tatetian** [comment]
I agree that we can make the implementation more robust even when the user programs are written in a "bad" way. We need a very light-weight hash implementation that is robust against common "bad" patterns of futex addresses. You give one example of such patterns, but I suppose there are other possibilities.

@tangruize Do you know what Linux does?

**Comment by tangruize** [comment]
Linux uses `jhash2` for key hashing:

- https://github.com/torvalds/linux/blob/7eef7e306d3c40a0c5b9ff6adc9b273cc894dbd5/kernel/futex/core.c#L117-L123
- https://github.com/torvalds/linux/blob/7eef7e306d3c40a0c5b9ff6adc9b273cc894dbd5/include/linux/jhash.h#L117-L146

**Comment by tatetian** [comment]
Fixed by https://github.com/asterinas/asterinas/pull/2339
